jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormData.java

```diff
@@ -552,28 +583,34 @@ public void onFailure(Throwable failure)
         private void fail(Throwable cause)
         {
             List<MultiPart.Part> partsToFail;
-            List<Content.Chunk> partChunksToFail;
             try (AutoLock ignored = lock.lock())
             {
                 if (failure != null)
                     return;
                 failure = cause;
-                partsToFail = new ArrayList<>(parts);
+                partsToFail = List.copyOf(parts);
                 parts.clear();
-                partChunksToFail = new ArrayList<>(partChunks);
+                partChunks.forEach(Content.Chunk::release);
                 partChunks.clear();
             }
             partsToFail.forEach(p -> p.fail(cause));
-            partChunksToFail.forEach(Retainable::release);
             close();
             delete();
         }
 
+        private SeekableByteChannel fileChannel()
+        {
+            try (AutoLock ignored = lock.lock())
+            {
+                return fileChannel;
+            }
+        }
+
         private void delete()
         {
-            try
+            try (AutoLock ignored = lock.lock())
             {
-                if (fileChannel != null)
+                if (filePath != null)
                     Files.delete(filePath);
```

Comment:
Can we do filesystem I/O outside locking?

Reply:
Done.